### ISSUE TITLE
company-cmake complete in string (pull request?)

### ISSUE THREAD

**Issue by oracleyue** [company-mode/company-mode#713]
(Since it is just one line to update, for simplicity, I present it here directly.)

** Problem Description
For example, in the following case as illustrated, we definitely want the company to complete it. Most functions in cmake could be memorized, while these variables are really nasty. We will expect company to help us. It makes little sense to complete using company-abbev etc. (The double quotes are mostly inserted in pairs, so we need the completion in strings.)

<img width="470" alt="screen shot 2017-09-27 at 15 35 38" src="https://user-images.githubusercontent.com/5046605/30916436-b6b1913a-a399-11e7-9edf-565095f1b0a7.png">

** Update on codes
Comment Line 190 in `company-cmake.el` or you might offer a variable similar to `company-dabbrev-code-everywhere`.

<img width="551" alt="screen shot 2017-09-27 at 15 32 16" src="https://user-images.githubusercontent.com/5046605/30916184-10d67ab4-a399-11e7-91be-8479f0535cc4.png">

Many thanks!

**Comment by dgutov** [comment]
Shouldn't it check that there is a `${` before the current prefix, though?

**Comment by oracleyue** [comment]
@dgutov Unfortunately, it doesn't work. I will check your codes to see if I can find the reason. (where I turned off `company-dabbrev-code-everywhere`.

<img width="643" alt="screen shot 2017-09-27 at 18 36 01" src="https://user-images.githubusercontent.com/5046605/30925987-fd2ca040-a3b3-11e7-9e83-a31da31c5f6b.png">
<img width="629" alt="screen shot 2017-09-27 at 18 35 37" src="https://user-images.githubusercontent.com/5046605/30925992-00eec550-a3b4-11e7-9c00-408bcdce3f2a.png">

**Comment by oracleyue** [comment]
@dgutov Does it work for you, in the two cases shown in your figures? (when you type double quotes and braces in or not in pairs)

**Comment by dgutov** [comment]
What doesn't work? You're not showing the new code.

**Comment by oracleyue** [comment]
@dgutov In you `company-cmake.el` (offered by the `company` package), I don't see any matching of  `${` or open an exception to start complete when there is `${` prefixed.  See the attached figure, it's `and`: the completion by `company-cmake` is turned off as long as `company-in-string-or-comment` returns `t`.

<img width="550" alt="screen shot 2017-09-27 at 18 51 52" src="https://user-images.githubusercontent.com/5046605/30926349-1a239f22-a3b5-11e7-9aa7-6a3b49d38093.png">

**Comment by dgutov** [comment]
Let me rephrase: simply removing this check is not good enough.

Instead, when inside a string, new code should check whether it is inside an interpolation. I'm not saying it works now.

**Comment by oracleyue** [comment]
FYI, it's `company 20170926.1515`, which is the latest version on `elpa`. Do you mean the version on github? Let me check.

**Comment by oracleyue** [comment]
@dgutov Sorry, maybe I am wrong. I didn't see it from your codes (also github version) that it makes an exception of `${` when `company-in-string-or-comment` returns `t`.

**Comment by oracleyue** [comment]
@dgutov Ah, got it. I agree. Make a function to check the prefix of `${` is definitely better.

**Comment by oracleyue** [comment]
@dgutov 
>What doesn't work? You're not showing the new code.

Sorry for misunderstanding your comments before. I am in my trip and didn't check carefully. I thought you were saying your latest codes have realized this check. I am working on it now. Let's see if I can make it.

**Comment by oracleyue** [comment]
Here is the function to test the prefix of `${`. I have soundly test it and it works well.
However, I never really program in elisp before. I just used the most basic functions to make it work. It moves cursors multiple times, but I just need the positions instead of really move there. I would like to make it neat and remove these unnecessary operations. Could you check it and give advices on refinement? Many thanks!
Once you approve on the coding style, I can file the pull request in the regular way to ease your update.


#### Codes update in `company-cmake.el`

Before the function `company-cmake()`, add the function:

```elisp
(defun company-cmake-prefox-dollar-brace-p ()
  "Test if the current char is prefixed with ${ in the current line."
  (setq-local position-current (point))
  (beginning-of-line)
  (setq-local position-beg-of-line (point))
  (end-of-line)
  (setq-local position-end-of-line (point))
  (goto-char position-current)

  (if (re-search-backward "\$\{" position-beg-of-line t)
      (progn
        (setq-local position-matched (point))
        (goto-char position-current))
    (setq-local position-matched nil))
  (if (re-search-backward "\}" position-beg-of-line t)
      (progn
        (setq-local position-matched-right-brace (point))
        (goto-char position-current))
    (setq-local position-matched-right-brace nil))

  (if (or (and position-matched
               position-matched-right-brace
               (> position-matched position-matched-right-brace))
          (and position-matched
               (not position-matched-right-brace)))
      t
    nil))
```

and the line of testing conditions in `company-cmake()`:
```elisp
(or (not (company-in-string-or-comment))
    (company-cmake-prefox-dollar-brace-p))
```

**Comment by oracleyue** [comment]
Move to the pull request: https://github.com/company-mode/company-mode/pull/714
